packages/monaco/src/browser/monaco-workspace.ts

```diff
@@ -159,7 +159,7 @@ export class MonacoWorkspace {
     protected readonly suppressedOpenIfDirty: MonacoEditorModel[] = [];
 
     protected openEditorIfDirty(model: MonacoEditorModel): void {
-        if (this.suppressedOpenIfDirty.indexOf(model) !== -1) {
+        if (model.suppressOpenEditorWhenDirty || this.suppressedOpenIfDirty.indexOf(model) !== -1) {
```

Comment:
I believe this is the change you are referring to with https://github.com/eclipse-theia/theia/pull/8910/files#r586280870, and the fact we should not use the application shell.

I'm trying to understand the code here, why do we want to omit opening dirty custom editors?
The `model.suppressOpenEditorWhenDirty` was added to the custom-editor model only to exclude this call which does not feel like the best solution, and is pretty hidden as an implementation.

Could we not do (or something along the lines):

```ts
if (model instanceof MonacoCustomEditorModel || ...`) { ... }
```

Given that we extend the `MonacoCustomEditorModel`?

Reply:
> I'm trying to understand the code here, why do we want to omit opening dirty custom editors?
> The `model.suppressOpenEditorWhenDirty` was added to the custom-editor model only to exclude this call which does not feel like the best solution, and is pretty hidden as an implementation.

We need to skip **openEditorIfDirty** logic because `MonacoEditor.findByDocument(this.editorManager, model).length === 0` will always result in true as `CustomEditWidget` (container of the custom model) is created using `plugin-custom-editor` factory id while `EditorWidget` is created using `code-editor-opener`, which would then result in creation of another model ref while the custom one exits already. 

I agree this isn't cleanest solution, if you can improve on it that would be great.

> Could we not do (or something along the lines):
> 
> ```ts
> if (model instanceof MonacoCustomEditorModel || ...`) { ... }
> ```

I cannot do **instanceof CustomTextEditorModel** check because that would require an import from `plugin-ext` which would result in a cyclic reference problem.

Reply:
@danarad05 thank you for explaining it further, I'd be fine to proceed with the way you're doing it as not to block the pull-request any longer, it will always be possible to refactor in the future anyways if we discover a cleaner solution.